### PR TITLE
Document Azure storage needs to set MEDIA_ROOT

### PR DESCRIPTION
**Pull request by fao89**
I've started a PoC for testing azure storage: https://github.com/pulp/pulp_ansible/pull/649
and found this issue
```
(pulp) [vagrant@pulp3-source-fedora34 pulp_ansible]$ nc 127.0.0.1 4444
> /home/vagrant/devel/pulpcore/pulpcore/app/tasks/importer.py(403)pulp_import()
-> if not default_storage.exists(dest):
(Pdb) dest
'/var/lib/pulp/media/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac'
(Pdb) default_storage.exists(dest)
*** django.core.exceptions.SuspiciousOperation: Attempted access to '/var/lib/pulp/media/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac' denied.
(Pdb) default_storage.list_all()
['pulp3/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac', 'pulp3/artifact/4e/042680c4cf9c54296e22e388e6058c469d9a09348523122d44165c2f2d61a8', 'pulp3/artifact/5a/9413fe085b128933884be1fb3f5094ac257e2383a0e869ce17b3c15624ae16', 'pulp3/artifact/6c/a52bdcb651b80b9655805bd44baf6e56c6b6de88210e1b412334be1278e4a5', 'pulp3/artifact/c1/05d0816dcfd2a4fa969ba063acdddf0fa43719fea8c556b2c70e96d60f84f2', 'pulp3/artifact/c1/e05337c06b438780bca6173925bc2f94904eb8cafb5595e2c7de2082009cb6', 'pulp3/pulp3/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac', 'pulp3/pulp3/artifact/4e/042680c4cf9c54296e22e388e6058c469d9a09348523122d44165c2f2d61a8', 'pulp3/pulp3/artifact/5a/9413fe085b128933884be1fb3f5094ac257e2383a0e869ce17b3c15624ae16', 'pulp3/pulp3/artifact/6c/a52bdcb651b80b9655805bd44baf6e56c6b6de88210e1b412334be1278e4a5', 'pulp3/pulp3/artifact/c1/05d0816dcfd2a4fa969ba063acdddf0fa43719fea8c556b2c70e96d60f84f2', 'pulp3/pulp3/artifact/c1/e05337c06b438780bca6173925bc2f94904eb8cafb5595e2c7de2082009cb6', 'pulp3/tmp/files/95/597048-d7e4-4268-9d6d-96a4d22d365d']
(Pdb) dest = os.path.join(settings.AZURE_LOCATION, base_path)
(Pdb) default_storage.exists(dest)
True
(Pdb) dest
'pulp3/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac'                                         
```

### CODE REVIEW

**Review by mdellweg — COMMENTED** [review]
I don't understand this change.
- Why is it only needed in this place? We also safe artifacts elsewhere.
- Why is it only needed for azure? Is the storage framework not abstracting enough?

**Comment by fao89** [comment]
> I don't understand this change.

I never used azure, so I was guided by pulp_ansible tests


> * Why is it only needed in this place? We also safe artifacts elsewhere.
> * Why is it only needed for azure? Is the storage framework not abstracting enough?

honestly, I don't know, I only tried to fix pulp_ansible import/export tests,
I don't know azure, but from what I understood the permission is restricted to AZURE_LOCATION, so you are denied when trying other paths. 

AZURE_LOCATION seems to be the equivalent to MEDIA_ROOT on the azure world,

AZURE_LOCATION: Default location for the uploaded files. This is a path that gets prepended to every file name.  - https://django-storages.readthedocs.io/en/latest/backends/azure.html

MEDIA_ROOT: Absolute filesystem path to the directory that will hold user-uploaded files. - https://docs.djangoproject.com/en/3.2/ref/settings/#std:setting-MEDIA_ROOT

**Comment by fao89** [comment]
> It seems strange that the path to the artifact in Azure starts with `/var/lib/pulp/media`. Is that where you are storing them in Azure? Also, what value do you have for MEDIA_ROOT? I know for S3, I typically use `''`.

it is the default value:
https://github.com/pulp/pulpcore/blob/master/pulpcore/app/settings.py#L38

pulpcore docs don't mention setting MEDIA_ROOT for azure:
https://docs.pulpproject.org/pulpcore/installation/storage.html#configuring-pulp-to-use-azure-blob-storage

here are the variables I set for azure:
https://github.com/fao89/pulp_ansible/blob/azure/.ci/ansible/settings.py.j2#L28-L37

**Comment by daviddavis** [comment]
Is there a way to confirm whether the artifacts in azure are stored under `/var/lib/pulp/media/`? If they are not, then I think that's a problem.

> it is the default value:
> https://github.com/pulp/pulpcore/blob/master/pulpcore/app/settings.py#L38

This doesn't sound correct to me. I believe it should be the same as AZURE_LOCATION.

> pulpcore docs don't mention setting MEDIA_ROOT for azure:
> https://docs.pulpproject.org/pulpcore/installation/storage.html#configuring-pulp-to-use-azure-blob-storage

I wouldn't rely to much on the docs. We've been accepting contributions without actually testing to make sure that they work.

**Comment by fao89** [comment]
> Is there a way to confirm whether the artifacts in azure are stored under `/var/lib/pulp/media/`? If they are not, then I think that's a problem.

I put the pdb at PR description, they are stored in `pulp3` which is `AZURE_LOCATION`

> I wouldn't rely to much on the docs. We've been accepting contributions without actually testing to make sure that they work.

I just mentioned the docs because instead of code we may need to change the docs

File: pulpcore/app/tasks/importer.py [hunk 1]
```diff
@@ -398,6 +398,8 @@ def validate_and_assemble(toc_filename):
                 base_path = os.path.join("artifact", artifact.sha256[0:2], artifact.sha256[2:])
                 src = os.path.join(temp_dir, base_path)
                 dest = os.path.join(settings.MEDIA_ROOT, base_path)
+                if getattr(settings, "AZURE_LOCATION", None):
+                    dest = os.path.join(settings.AZURE_LOCATION, base_path)
```
Comment (fao89) [inline pulpcore/app/tasks/importer.py:402]:
```
(pulp) [vagrant@pulp3-source-fedora34 pulp_ansible]$ nc 127.0.0.1 4444
> /home/vagrant/devel/pulpcore/pulpcore/app/tasks/importer.py(403)pulp_import()
-> if not default_storage.exists(dest):
(Pdb) dest
'/var/lib/pulp/media/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac'
(Pdb) default_storage.exists(dest)
*** django.core.exceptions.SuspiciousOperation: Attempted access to '/var/lib/pulp/media/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac' denied.
(Pdb) default_storage.list_all()
['pulp3/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac', 'pulp3/artifact/4e/042680c4cf9c54296e22e388e6058c469d9a09348523122d44165c2f2d61a8', 'pulp3/artifact/5a/9413fe085b128933884be1fb3f5094ac257e2383a0e869ce17b3c15624ae16', 'pulp3/artifact/6c/a52bdcb651b80b9655805bd44baf6e56c6b6de88210e1b412334be1278e4a5', 'pulp3/artifact/c1/05d0816dcfd2a4fa969ba063acdddf0fa43719fea8c556b2c70e96d60f84f2', 'pulp3/artifact/c1/e05337c06b438780bca6173925bc2f94904eb8cafb5595e2c7de2082009cb6', 'pulp3/pulp3/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac', 'pulp3/pulp3/artifact/4e/042680c4cf9c54296e22e388e6058c469d9a09348523122d44165c2f2d61a8', 'pulp3/pulp3/artifact/5a/9413fe085b128933884be1fb3f5094ac257e2383a0e869ce17b3c15624ae16', 'pulp3/pulp3/artifact/6c/a52bdcb651b80b9655805bd44baf6e56c6b6de88210e1b412334be1278e4a5', 'pulp3/pulp3/artifact/c1/05d0816dcfd2a4fa969ba063acdddf0fa43719fea8c556b2c70e96d60f84f2', 'pulp3/pulp3/artifact/c1/e05337c06b438780bca6173925bc2f94904eb8cafb5595e2c7de2082009cb6', 'pulp3/tmp/files/95/597048-d7e4-4268-9d6d-96a4d22d365d']
(Pdb) dest = os.path.join(settings.AZURE_LOCATION, base_path)
(Pdb) default_storage.exists(dest)
True
(Pdb) dest
'pulp3/artifact/36/0548ba80e3dce478b7915ca89a5613dc80c650a55b96f5491012a8297e12ac'                                         
```

**Review by mdellweg — APPROVED** [review]
This sounds much more like it!

**Comment by daviddavis** [comment]
Looks good. 👍 
 
I think the commit message needs update.

**Review by daviddavis — APPROVED** [review]

**Pull request merged**